proof/refine/Invariants_H.thy

```diff
@@ -1390,27 +1150,10 @@ lemma obj_atI' [intro?]:
   \<Longrightarrow> obj_at' P ptr s"
   unfolding obj_at'_def by (auto)
 
-lemma vcpu_at_is_vcpu':
-  "vcpu_at' v = ko_wp_at' is_vcpu' v"
-  apply (rule ext)
-  apply (clarsimp simp: typ_at'_def is_vcpu'_def ko_wp_at'_def)
-  apply (rule iffI; clarsimp?)
-  apply (case_tac ko; simp; rename_tac ako; case_tac ako; simp)
-  done
-
 lemma cte_at'_def:
   "cte_at' p s \<equiv> \<exists>cte::cte. fst (getObject p s) = {(cte,s)}"
   by (simp add: cte_wp_at'_def)
 
-
-lemma tcb_cte_cases_simps[simp]:
-  "tcb_cte_cases 0  = Some (tcbCTable, tcbCTable_update)"
-  "tcb_cte_cases 32 = Some (tcbVTable, tcbVTable_update)"
-  "tcb_cte_cases 64 = Some (tcbReply, tcbReply_update)"
-  "tcb_cte_cases 96 = Some (tcbCaller, tcbCaller_update)"
-  "tcb_cte_cases 128 = Some (tcbIPCBufferFrame, tcbIPCBufferFrame_update)"
```

Comment:
With the numbers unfolded, these would be arch specific (or at least mode/machine word size specific). If we could avoid unfolding `cteLevelBits` in the proofs, then the lemma would stay generic or would not even be needed. Not sure it's worth the hassle at this point, but might be worth a FIXME.

Reply:
Thinking more about it, the main reason why we tend to unfold these is because it is automatic for Isabelle to decide that e.g. `0 != 32`, but it can't automatically decide that `0 * cteLevelBits != 1 * cteLevelBits`. There is a mechanism somewhere for turning a concrete distinct list into a set of inequality `simp` lemmas. We could use that here, and prove `distinct [0 * cteLevelBits, 1 * cteLevelBits, ...]` once and use that. Then Isabelle should be able to do symbolically what it now needs the numbers for. I'll try find what that tool is called.

Reply:
Right, it's not a tool, and it's not merged yet, it's just an application of `simplified`. Here is an example: https://github.com/seL4/l4v/blob/dc2fd26954724fca79d175528726314de834605a/proof/crefine/ARM/CSpace_C.thy#L3064-L3088

Reply:
I've deployed something, it'll need review once I push it

Reply:
Please see the eventual infrastructure and the FIXMEs about overlapping lemmas. It follows on from our discussion on MM about being unable to sort the conjuncts and remove the redundant ones, and I still don't know how to do that.

Reply:
:+1: Happy with that for now. To make it nice, we probably have to adjust some of these TCB/CTE proofs quite a bit, which may even reach into spec changes. Something for some time in the future, but not for the current arch split pass.